### ISSUE TITLE
Do not ping MySQL on connection checkout

### ISSUE THREAD

**Issue by finbarrtimbers** [finbarrtimbers/rails#31]
Now on every single connection checkout we ping MySQL with
`mysql_ping`. This is nothing for an app with a couple app servers, but in case of Shopify we have over 500 servers, each running 4 containers with 12 unicorn processes. It means that on the first request after the deploy, we checkout 500 * 4 * 12 = **24000** connections. This creates a peak of 24K QPS on the master MySQL database and makes our data ops sad.

**Why is this bad**? Because these pings are not free. On every request we (usually) checkout two connections: 1) the shard, and 2) the master shard. This is two network roundtrips of roughly 1ms each. In addition, these calls are not free on the database either. Each call takes 100μs, whereas a small, indexed query takes 500μs. Given that we have such a staggering number of them, especially on the master shard which is checked out on connections to all shards, this adds up to an absurd amount of CPU time spent in Rails pinging connections and on the database processing those pings.

**Why is the ping there**? To health-check the connection and reconnect if it's not healthy. This is problematic though, because it's racy. If you perform the ping, and reconnect, by the time you hit the first query your connection might have been lost. For the majority of production, this is a no-op ping. It's legitimate in cases where the connection has gone stale (even that is rare). If we blindly remove the ping, in those cases we'd lose the entire request. This is not unreasonable, but this patch will ensure that even requests from stale or closed connections will succeed by retrying the first query after a connection checkout.

**This is not only the Shopify problem**. Same issue have been described by [Percona](https://www.percona.com/blog/2010/05/05/checking-for-a-live-database-connection-considered-harmful/) and [by Groupon on their Rails patch](https://engineering.groupon.com/2013/rails/removing-the-ping-sting-from-rails-db-connections/). As outlined in Percona blog post, 73% of the server’s load was consumed by checking to see if the connection is still alive.

**Summary:** while the code behind my implementation is not ideal, I think that many large-scale Rails apps such as Basecamp will benefit from the option to skip MySQL ping. I'm open to the discussion how we can make it an optional thing.

All credits for discovering the issue and the patch belong to @Sirupsen and @rafaelfranca.

@tenderlove @arthurnn @sgrif @jeremy

**Comment by arthurnn** [comment]
Is this a new issue? if so, it should be open on the rails/rails repo, so more folks will be able to jump in.

**Comment by finbarrtimbers** [comment]
No, sorry. I'm trying to test some issue management software that I have, so I copied the issues for testing. If that's a problem I can delete this repo.

**Comment by arthurnn** [comment]
@finbarrtimbers np
